crates/forge/bin/cmd/clone.rs

```diff
@@ -407,74 +414,148 @@ fn update_config_by_metadata(
     // get config
     let path_config = ProjectPathsConfig::builder().build_with_root(root);
     // we will canonicalize the sources directory later
     let src_dir = &path_config.sources;
+    let lib_dir = &path_config.libraries[0];
     let contract_name = &meta.contract_name;
     let source_tree = meta.source_tree();
 
     // then we move the sources to the correct directories
     // we will first load existing remappings if necessary
     //  make sure this happens before dumping sources
     let mut remappings: Vec<Remapping> = Remapping::find_many(root);
-    // we also load the original remappings from the metadata
-    remappings.extend(meta.settings()?.remappings);
 
     // first we dump the sources to a temporary directory
     let tmp_dump_dir = root.join("raw_sources");
     source_tree
         .write_to(&tmp_dump_dir)
         .map_err(|e| eyre::eyre!("failed to dump sources: {}", e))?;
 
-    // check whether we need to strip the `contract` or `src` directories in the original sources.
-    // They are the source directories in foundry or hardhat projects. We do not want to preserve
-    // them in the cloned project.
-    // If there is any other directory other than `src` `contracts` or `lib`, we should not strip.
-    let strip_old_src = std::fs::read_dir(tmp_dump_dir.join(contract_name))?.all(|e| {
-        let Ok(e) = e else { return false };
-        let folder_name = e.file_name().to_string_lossy().to_string();
-        ["contracts", "src", "lib"].contains(&folder_name.as_str())
-    });
-    // move contract sources to the `src` directory
+    // check whether we need to re-organize directories in the original sources, since we do not
+    // want to put all the sources in the `src` directory if the original directory structure is
+    // well organized, e.g., a standard foundry project containing `src` and `lib`
+    //
+    // * if the user wants to keep the original directory structure, we should not re-organize.
+    // * if there is any other directory other than `src`, `contracts`, `lib`, `hardhat`,
+    //   `forge-std`,
+    // or not started with `@`, we should not re-organize.
+    let to_reorg = !no_reorg &&
+        std::fs::read_dir(tmp_dump_dir.join(contract_name))?.all(|e| {
+            let Ok(e) = e else { return false };
+            let folder_name = e.file_name();
+            folder_name == "src" ||
+                folder_name == "lib" ||
+                folder_name == "contracts" ||
+                folder_name == "hardhat" ||
+                folder_name == "forge-std" ||
+                folder_name.to_string_lossy().starts_with('@')
+        });
+
+    // ensure `src` and `lib` directories exist
+    eyre::ensure!(std::fs::metadata(root.join(src_dir)).is_ok(), "source directory must exists");
+    eyre::ensure!(std::fs::metadata(root.join(lib_dir)).is_ok(), "source directory must exists");
```

Comment:
Any reason not to use `Path::exists` instead of `metadata.is_ok()`? The former is more readable.

Reply:
Oh, yes, `Path::exists` is much better. I just copied-and-pasted my previous code and forgot this. Will fix.